source/opt/loop_fusion.cpp

```diff
@@ -0,0 +1,735 @@
+// Copyright (c) 2018 Google LLC.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "opt/loop_fusion.h"
+
+#include <algorithm>
+#include <vector>
+
+#include "opt/ir_context.h"
+#include "opt/loop_dependence.h"
+#include "opt/loop_descriptor.h"
+
+namespace spvtools {
+namespace opt {
+
+namespace {
+
+// Append all the loops nested in |loop| to |loops|.
+void CollectChildren(ir::Loop* loop, std::vector<const ir::Loop*>* loops) {
+  for (auto child : *loop) {
+    loops->push_back(child);
+    if (child->NumImmediateChildren() != 0) {
+      CollectChildren(child, loops);
+    }
+  }
+}
+
+// Return the set of locations accessed by |stores| and |loads|.
+std::set<ir::Instruction*> GetLocationsAccessed(
+    const std::map<ir::Instruction*, std::vector<ir::Instruction*>>& stores,
+    const std::map<ir::Instruction*, std::vector<ir::Instruction*>>& loads) {
+  std::set<ir::Instruction*> locations{};
+
+  for (const auto& kv : stores) {
+    locations.insert(std::get<0>(kv));
+  }
+
+  for (const auto& kv : loads) {
+    locations.insert(std::get<0>(kv));
+  }
+
+  return locations;
+}
+
+// Append all dependences from |sources| to |destinations| to |dependences|.
+void GetDependences(std::vector<DistanceVector>* dependences,
+                    LoopDependenceAnalysis* analysis,
+                    const std::vector<ir::Instruction*>& sources,
+                    const std::vector<ir::Instruction*>& destinations,
+                    size_t num_entries) {
+  for (auto source : sources) {
+    for (auto destination : destinations) {
+      DistanceVector dist(num_entries);
+      if (!analysis->GetDependence(source, destination, &dist)) {
+        dependences->push_back(dist);
+      }
+    }
+  }
+}
+
+// Apped all instructions in |block| to |instructions|.
+void AddInstructionsInBlock(std::vector<ir::Instruction*>* instructions,
+                            ir::BasicBlock* block) {
+  for (auto& inst : *block) {
+    instructions->push_back(&inst);
+  }
+
+  instructions->push_back(block->GetLabelInst());
+}
+
+}  // namespace
+
+bool LoopFusion::UsedInContinueOrConditionBlock(
+    ir::Instruction* phi_instruction, ir::Loop* loop) {
+  auto condition_block = loop->FindConditionBlock()->id();
+  auto latch_block = loop->GetLatchBlock()->id();
```

Comment:
As in other PRs, the continue target for a loop is not always the same as the latch.  The name is a little confusing.

Reply:
`GetLatchBlock` actually returns the continue block (https://github.com/KhronosGroup/SPIRV-Tools/blob/master/source/opt/loop_descriptor.h#L88). It should probably be renamed in a separate pull request.